### PR TITLE
[Refactor] Add infra-driver, cluster-runtime, container-runtime, registry-configurator interface

### PR DESCRIPTION
**Pull request by justadogistaken**

### CODE REVIEW

**Comment by justadogistaken** [comment]
@VinceCui FYI

File: pkg/container-runtime/installer.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	//Rollback() (ContainerRuntimeInfo, error)
+}
+
+func NewInstaller(conf Config) Installer {
+
+}
+
+type Config struct {
+	Type         string
+	LimitNofile  string `json:"limitNofile,omitempty" yaml:"limitNofile,omitempty"`
```
Comment (kakaZhou719) [inline pkg/container-runtime/installer.go:29]: should this field be `int` type? and  Is this field for checking OS configuration?

Comment (kakaZhou719) [inline pkg/container-runtime/installer.go:29]: should this field be `int` type?

**Comment by allencloud** [comment]
@sealerio/sealer-reviewers 
I still have several questions for this refactoring. I must say again that I am the one who totally support refactoring sealer's arch. While to achieve this goal, I wish the refactoring team would stay serious to face the challenges, back to basic.

1. I see this pull request already contains two much code change without maintainers' review, maybe contains reviews just from @kakaZhou719 and @VinceCui . Then, how to guarantee that this pull request could be rebased to the latest main branch? The rebase work including solving conflict would consume so much energy of the entire team.
2. How to detect whether the rebase work would ignore the original bugfix code by accident? The refactoring will indeed change the architecture, but for the detailed functionality, we have no way to notice the side effect.
3. No testing is guaranteed, how to prove the quality of refactoring? Are we changing the legacy bad world into a new bad world.  How to reduce the possibility that we refatored the arch into another unsuccessful arch.

In a word, if there is no a reasonably detailed way for this pull request to guarantee merging to the main branch, the refactoring seems to help less.

File: pkg/runtime/kubernetes/runtime.go [hunk 2]
```diff
@@ -15,390 +15,244 @@
...
 	k := &Runtime{
-		cluster: cluster,
+		infra: infra,
 		Config: &Config{
-			ClusterFileKubeConfig: clusterFileKubeConfig,
-			APIServerDomain:       DefaultAPIserverDomain,
+			KubeadmConfigFromClusterFile: clusterFileKubeConfig,
+			APIServerDomain:              DefaultAPIserverDomain,
+			//TODO
+			LvsImage:             fmt.Sprintf("%s/fanux/lvscare:latest", registryInfo.URL),
```
Comment (VinceCui) [inline pkg/runtime/kubernetes/runtime.go:65]: use labring/lvscare:v1.1.3-beta.8

File: pkg/runtime/kubernetes/init.go [hunk 3]
```diff
@@ -17,170 +17,110 @@ package kubernetes
...
-		&k.KubeProxyConfiguration)
-}
 
-func (k *Runtime) handleKubeadmConfig() {
-	//The configuration set here does not require merge
-	k.setInitAdvertiseAddress(k.cluster.GetMaster0IP())
-	k.setControlPlaneEndpoint(fmt.Sprintf("%s:6443", k.getAPIServerDomain()))
-	if k.APIServer.ExtraArgs == nil {
-		k.APIServer.ExtraArgs = make(map[string]string)
+	//TODO, save it into kubernetes
```
Comment (VinceCui) [inline pkg/runtime/kubernetes/init.go:65]: no need do this TODO, kubeadm will do it by itself.

**Review by VinceCui — COMMENTED** [review]
There are so many reviews so far, I will continue to add

**Pull request merged**